### DISCUSSION TITLE
SO32 setting not retained after restart

### DISCUSSION THREAD

**Discussion by Malt0178** [arendst/Tasmota#15088] [Q&A] [answered]
Is there a method to force settings to be retained? Martin Jerry SD01s configured as PWM Dimmers with SO32 5 revert to the default (40) after a restart 1. I understand the default is to save configuration updates 1 sec after a change. For me, this works in 9.1.0 and fails in both 10.1.0 and 11.0.0. Interestingly, the SO32 setting is retained with 10.1.0 in the Martin Jerry Non-dimming switches.

**Comment by ascillato** [comment, marked as answer]
SO32 is retained. Please check the logs because it seems that there is a problem in your configuration that forces Tasmota to go to defaults due to boot loop.

**Reply by Malt0178** [reply]
Thanks for the reply. I'm not seeing anything unusual in the logs or the reversion of any other settings to their defaults. I did a little more checking on an 11.0.0 dimmer and I can set SO32 to 1-4 or 6-10 and the values are all retained after restarts. Each time setting SO32 to 5 followed by a restart 1 causes the setting to revert to 40. The console log does indicate the changed values are being saved to flash.

**Reply by ascillato** [reply]
I have just tested on latest version and `SO32 5` remains in flash. Is not returned to default.

Please set weblog to 4 and do that test again. You should see in the log (at the beginning) that some settings are being set to default.

**Reply by Malt0178** [reply]
Retested with Weblog 4 (was on in the previous tests). The results are the same. In comparing the logs they are effectively identical. No indication of defaults being set that I see. Failed log after SO32 5 and restart 1
[Failed Case Log.txt](https://github.com/arendst/Tasmota/files/8218355/Failed.Case.Log.txt)
.

**Comment by barbudor** [comment]
Your log is missing everything that happens betwen the boot start and the MQTT connection
Would be even better to get everything from the initial SO32 5 / Restart 1 until the reboot and your final SO32 command

Did you ever sent the SO32 command to Tasmota through MQTT from either H.A. or another software ? You could have sent once a SO32 40 command retained and this remains stored in the MQTT broker. Tasmota would then see that at every restart.
Do you have a MQTT tool such as MQTT Explorer ? That would help identifying retained messages.

**Reply by barbudor** [reply]
I can confirm that I can see the problem on a nodemcu configured like a MJ

```
23:00:17.310 CMD: so32 5
23:00:17.317 MQT: stat/dev-4119/SO = {"SetOption32":5}
23:00:22.370 CMD: restart 1
23:00:22.376 MQT: stat/dev-4119/RESTART = {"Restart":"Restarting"}

00:00:00.002 HDW: ESP8266EX
00:00:00.054 CFG: Loaded from flash at FB, Count 15
00:00:00.066 Project tasmota - dev-4119 Version 11.0.0.3(bhome)-2_7_4_9(2022-03-09T02:01:40)
00:00:04.553 WIF: Connecting to AP1 BNETWLAN Channel 1 BSSId 88:C3:97:F2:25:EC in mode 11n as dev-4119...
00:00:06.531 WIF: Connected
00:00:06.784 HTP: Web server active on dev-4119 with IP address 192.168.168.199
23:00:32.386 MQT: Attempting connection...
23:00:32.396 MQT: Connected
23:00:32.399 MQT: tele/dev-4119/LWT = Online (retained)
23:00:32.401 MQT: cmnd/dev-4119/POWER = 
23:00:32.413 MQT: tele/dev-4119/INFO1 = {"Info1":{"Module":"MJ-SD01 Dimmer","Version":"11.0.0.3(bhome)","FallbackTopic":"cmnd/3C71BF2AD017_fb/","GroupTopic":"cmnd/alldevices/"}}
23:00:32.429 MQT: tele/dev-4119/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"dev-4119","IPAddress":"192.168.168.199"}}
23:00:32.440 MQT: tele/dev-4119/INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":5}}
23:00:32.450 MQT: stat/dev-4119/POWER = {"POWER":"ON"}
23:00:32.454 MQT: stat/dev-4119/POWER = ON (retained)
23:00:37.380 MQT: tele/dev-4119/STATE = {"Time":"2022-03-09T23:00:37+01:00","Uptime":"0T00:00:14","UptimeSec":14,"Heap":29,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"OFF","Dimmer":10,"Fade":"OFF","Speed":1,"LedTable":"ON","Wifi":{"AP":1,"SSId":"XXXXX","BSSId":"88:C3:97:F2:25:EC","Channel":1,"Mode":"11n","RSSI":100,"Signal":-48,"LinkCount":1,"Downtime":"0T00:00:08"}}
23:00:42.329 CMD: so32
23:00:42.336 MQT: stat/dev-4119/SO = {"SetOption32":40}
```

**Comment by barbudor** [comment]
This is by design: 5 => 40
https://github.com/arendst/Tasmota/blob/a048e6bdfd936e2dc4a23df19b59c26766132f7f/tasmota/xdrv_35_pwm_dimmer.ino#L101-L105

This seems to be legacy and is probably not needed anymore.
At least you can use 4 or 6 which should impact user experience

@arendst Do you think the above code is still needed ? This seems to be a legacy from old values.
If still needed, it would probably be better located in the setttings upgrade.

Hi @pcdiem this is your art : https://github.com/arendst/Tasmota/commit/f075a89d6b72030b2e40cd3c97a74090665c5820
Is that to manage a setting upgrade from an old version ?

**Reply by Malt0178** [reply]
Thanks for the help figuring this out. 6 seems like a good value.

**Reply by ascillato** [reply]
Good catch! I couldn't reproduce it in my previous attempt because I haven't defined USE_PWM_DIMMER. Now I can reproduce it.

**Reply by arendst** [reply]
I suggest to remove the code.
